onadata/apps/logger/data_migration/restore_backup.py

```diff
@@ -33,19 +32,67 @@ def restore_xform_backup(self):
                 self._restore_survey_from_backup(survey, backup_survey)
 
     def migrate_survey_to_prev_schema(self, survey):
-        changes = self.xform_backup.changes
-        reversed_decisions = self.get_reversed_migration_decisions(changes)
+        changes_list = self._get_migration_changes_in_between(self.xform, self.xform_backup)
+        decisions = MigrationDecisioner.convert_changes_to_decisions(
+            self._merge_migration_changes(changes_list)
+        )
         data_migrator_factory(
             self.xform_backup, self.xform,
-            backup_data=False, **reversed_decisions
+            backup_data=False, **decisions
         ).migrate()
 
+    @classmethod
+    def _merge_migration_changes(cls, changes_list):
+        empty_changes = MigrationDecisioner.construct_changes()
+        return cls._merge_migration_changes_aux(changes_list, empty_changes)
+
+    @classmethod
+    def _merge_migration_changes_aux(cls, changes_list, merged_changes):
+        if not len(changes_list):
+            return merged_changes
+        changes = changes_list.pop(0)
+        new_merged_changes = cls._merge_changes(merged_changes, changes)
+        return cls._merge_migration_changes_aux(changes_list, new_merged_changes)
+
+    @classmethod
+    def _merge_changes(cls, curr_changes, new_changes):
+        merge_and_remove_duplicates = lambda x1, x2: list(set(x1) | set(x2))
+
+        return MigrationDecisioner.construct_changes(
+            new=merge_and_remove_duplicates(curr_changes[NEW_FIELDS_KEY],
+                                            new_changes[NEW_FIELDS_KEY]),
+            removed=merge_and_remove_duplicates(curr_changes[RM_FIELDS_KEY],
+                                                new_changes[RM_FIELDS_KEY]),
+            modified=cls._transitive_merge(curr_changes[MOD_FIELDS_KEY],
+                                           new_changes[MOD_FIELDS_KEY]),
+        )
+
     @staticmethod
-    def get_reversed_migration_decisions(changes):
-        reversed_changes = MigrationDecisioner.reverse_changes(changes)
-        return MigrationDecisioner.convert_changes_to_decisions(reversed_changes)
+    def _transitive_merge(curr_changes, new_changes):
+        curr_changes = {
+            new_var_name: new_changes.pop(prev_var_name, prev_var_name)
+            for new_var_name, prev_var_name in curr_changes.items()
+        }
+        curr_changes.update(new_changes)
+        return curr_changes
 
-    def _get_xform_backup(self, version, restore_last):
+    @staticmethod
+    def _get_path_in_version_tree(xform, backup):
+        return VersionTree.objects.find_path(xform.version_tree, backup.version_tree)
+
+    @classmethod
+    def _get_migration_changes_in_between(cls, xform, backup):
+        get_migration_changes = lambda vt: vt.version.migration_changes
```

Comment:
maybe def get_path_migration_changes(path) instead of lambda + using map twice below?